js/src/builtin/streams/ReadableStreamBYOBReader.cpp

```diff
@@ -6,43 +6,552 @@
+#include "vm/NativeObject-inl.h"  // js::ThrowIfNotConstructing
+
+using JS::CallArgs;
 using JS::Handle;
+using JS::Rooted;
+using JS::RootedObject;
+using JS::RootedValueArray;
+using JS::Value;
+
+using js::GetErrorMessage;
+using js::ListObject;
+using js::NewObjectWithClassProto;
+using js::PromiseObject;
+using js::PullIntoDescriptor;
+using js::ReadableByteStreamController;
+using js::ReadableStream;
+using js::ReadableStreamBYOBReader;
+using js::ReadableStreamReader;
+using js::UnwrapAndTypeCheckThis;
 
 /*** 3.7. Class ReadableStreamBYOBReader *********************************/
 
 /**
  * Stream spec, 3.7.3. new ReadableStreamBYOBReader ( stream )
  * Steps 2-5.
  */
-[[nodiscard]] JSObject* js::CreateReadableStreamBYOBReader(
+ReadableStreamBYOBReader* js::CreateReadableStreamBYOBReader(
     JSContext* cx, Handle<ReadableStream*> unwrappedStream,
     ForAuthorCodeBool forAuthorCode, Handle<JSObject*> proto /* = nullptr */) {
+  Rooted<ReadableStreamBYOBReader*> reader(
+      cx, NewObjectWithClassProto<ReadableStreamBYOBReader>(cx, proto));
+  if (!reader) {
+    return nullptr;
+  }
+
+  // Step 1: If ! IsReadableStreamLocked(stream) is true, throw a TypeError
+  //         exception.
+  if (unwrappedStream->locked()) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_READABLESTREAM_LOCKED);
+    return nullptr;
+  }
+
   // Step 2: If ! IsReadableByteStreamController(
   //                  stream.[[readableStreamController]]) is false, throw a
   //         TypeError exception.
-  // We don't implement byte stream controllers yet, so always throw here.  Note
-  // that JSMSG_READABLESTREAM_BYTES_TYPE_NOT_IMPLEMENTED can't be used here
-  // because it's a RangeError (and sadly wpt actually tests this and we have a
-  // spurious failure if we don't make this a TypeError).
-  JS_ReportErrorNumberASCII(
-      cx, GetErrorMessage, nullptr,
-      JSMSG_READABLESTREAM_BYOB_READER_FOR_NON_BYTE_STREAM);
-
-  // Step 3: If ! IsReadableStreamLocked(stream) is true, throw a TypeError
+  if (!unwrappedStream->controller()->is<js::ReadableByteStreamController>()) {
+    JS_ReportErrorNumberASCII(
+        cx, GetErrorMessage, nullptr,
+        JSMSG_READABLESTREAM_BYOB_READER_FOR_NON_BYTE_STREAM);
+    return nullptr;
+  }
+
+  // Step 3: Perform ! ReadableStreamReaderGenericInitialize(this, stream).
+  // Step 4: Set this.[[readRequests]] to a new empty List.
+  if (!ReadableStreamReaderGenericInitialize(cx, reader, unwrappedStream,
+                                             forAuthorCode)) {
+    return nullptr;
+  }
+
+  return reader;
+}
+
+/**
+ * Stream spec, 3.6.3. new ReadableStreamBYOBReader ( stream )
+ */
+bool ReadableStreamBYOBReader::constructor(JSContext* cx, unsigned argc,
+                                           Value* vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+
+  if (!ThrowIfNotConstructing(cx, args, "ReadableStreamBYOBReader")) {
+    return false;
+  }
```

Comment:
I think we also need to check the argument has been supplied, so that we can return a specific TypeError message along the lines of `ReadableStreamBYOBReader constructor: At least 1 argument required, but only 0 passed`
```
if (!args.requireAtLeast(cx, "ReadableStreamBYOBReader constructor", 1)) {
    return false;
}
```

Reply:
Again, the spec doesn't check arg counts, and in this case doesn't even strictly validate it's a stream. But yes we should be validating that argument, and that's what the browser and Node.js seem to do.

Reply:
The spec defines it as a required argument via the webidl definition and not within any specific step - https://streams.spec.whatwg.org/#ref-for-default-reader-constructor

Reply:
This helper function - `UnwrapAndTypeCheckArgument` just below already handles the correct TypeError throw if the arg is not a readable stream.